server/src/main/java/org/elasticsearch/search/NestedDocuments.java

```diff
@@ -0,0 +1,182 @@
+     */
+    public boolean hasNonNestedParent(String path) {
+        return mapperService.documentMapper().hasNonNestedParent(path);
+    }
+
+    private class HasNestedDocuments implements LeafNestedDocuments {
+
+        final LeafReaderContext ctx;
+        final BitSet parentFilter;
+        final Map<String, BitSet> objectFilters = new HashMap<>();
+
+        int doc = -1;
+        int rootDoc = -1;
+        SearchHit.NestedIdentity nestedIdentity = null;
+
+        private HasNestedDocuments(LeafReaderContext ctx) throws IOException {
+            this.ctx = ctx;
+            this.parentFilter = parentDocumentFilter.getBitSet(ctx);
+            for (Map.Entry<String, BitSetProducer> filter : parentObjectFilters.entrySet()) {
+                BitSet bits = filter.getValue().getBitSet(ctx);
+                if (bits != null) {
+                    objectFilters.put(filter.getKey(), bits);
+                }
+            }
+        }
+
+        @Override
+        public SearchHit.NestedIdentity advance(int doc) {
+            assert doc >= 0 && doc < ctx.reader().maxDoc();
+            if (parentFilter.get(doc)) {
+                // parent doc, no nested identity
+                this.nestedIdentity = null;
+                this.doc = doc;
+                this.rootDoc = doc;
+                return null;
+            } else {
+                this.doc = doc;
+                this.rootDoc = parentFilter.nextSetBit(doc);
+                return this.nestedIdentity = loadNestedIdentity();
+            }
+        }
+
+        @Override
+        public int doc() {
+            assert doc != -1 : "Called doc() when unpositioned";
+            return doc;
+        }
+
+        @Override
+        public int rootDoc() {
+            assert doc != -1 : "Called rootDoc() when unpositioned";
+            return rootDoc;
+        }
+
+        @Override
+        public SearchHit.NestedIdentity nestedIdentity() {
+            assert doc != -1 : "Called nestedIdentity() when unpositioned";
+            return nestedIdentity;
+        }
+
+        private SearchHit.NestedIdentity loadNestedIdentity() {
+            String path = null;
+            for (Map.Entry<String, BitSet> objectFilter : objectFilters.entrySet()) {
+                if (objectFilter.getValue().get(doc)) {
+                    if (path == null || path.length() > objectFilter.getKey().length()) {
+                        path = objectFilter.getKey();
+                    }
+                }
+            }
+            if (path == null) {
+                throw new IllegalStateException("Cannot find object path for document " + doc);
+            }
+            SearchHit.NestedIdentity ni = null;
+            int currentLevelDoc = doc;
+            int parentNameLength;
+            while (path != null) {
+                String parent = mapperService.documentMapper().getNestedParent(path);
+                BitSet childBitSet = objectFilters.get(path);
```

Comment:
We don't really need a bitset here ? The child filter always move forward so we can build the `Scorer` instead and save some memory. That's what the current impl does so I'd prefer that we keep this logic.